### PR TITLE
Issue a warning when final field is overwritten #414

### PR DESCRIPTION
**Pull request by alamar**
#414

### CODE REVIEW

File: src/main/java/net/openhft/chronicle/wire/WireMarshaller.java [hunk 1]
```diff
@@ -461,6 +462,10 @@ public boolean isLeaf() {
             offset = unsafeObjectFieldOffset(field);
             key = field::getName;
             this.isLeaf = isLeaf;
+
+            if ((field.getModifiers() & Modifier.FINAL) != 0)
```
Comment (minborg) [inline src/main/java/net/openhft/chronicle/wire/WireMarshaller.java:466]: Suggest: `Modifier.isFinal(field.getModifiers))`

File: src/main/java/net/openhft/chronicle/wire/WireMarshaller.java [hunk 2]
```diff
@@ -450,6 +450,7 @@ public boolean isLeaf() {
 
         Comment commentAnnotation;
         Boolean isLeaf;
+        boolean isFinalNoWarning;
```
Comment (minborg) [inline src/main/java/net/openhft/chronicle/wire/WireMarshaller.java:453]: `final`

Comment (minborg) [reply src/main/java/net/openhft/chronicle/wire/WireMarshaller.java:453]: Ahh. ok. You change it after one warning. So no `final`.

File: src/main/java/net/openhft/chronicle/wire/WireMarshaller.java (hunk 1 above)
Comment (alamar) [reply src/main/java/net/openhft/chronicle/wire/WireMarshaller.java:466]: We seem to always use the form I've used, can we stick with it for consistency?

Comment (minborg) [reply src/main/java/net/openhft/chronicle/wire/WireMarshaller.java:466]: Yes

**Comment by alamar** [comment]
Turns out we have more than 10 tests which do exactly that - overwrite final fields.

What should we do - should we update them all?

@rogersimmons @RobAustin

**Comment by JerryShea** [comment]
> What should we do - should we update them all?
> 
IMO yes we should - our code should set a good example

**Comment by rogersimmons** [comment]
+1. Yes, please fix any tests which overwrite final fields.

File: src/main/java/net/openhft/chronicle/wire/WireMarshaller.java [hunk 3]
```diff
@@ -641,6 +650,14 @@ protected void readValue(Object o, Object defaults, ValueIn read, boolean overwr
             }
         }
 
+        protected void triggerFinalWarning() {
```
Comment (minborg) [inline src/main/java/net/openhft/chronicle/wire/WireMarshaller.java:653]: Suggest `assertNonFinal()`. In this version, there will be a warning but in later, an Exception would be thrown.

File: src/test/java/net/openhft/chronicle/wire/NestedMapsTest.java [hunk 4]
```diff
@@ -255,9 +255,9 @@ public void testMapReadAndWrite() {
     }
 
     static class Mapped extends SelfDescribingMarshallable {
-        final Set<String> words = new LinkedHashSet<>();
-        final List<Integer> numbers = new ArrayList<>();
-        final Map<String, String> map1 = new LinkedHashMap<>();
-        final Map<String, Double> map2 = new LinkedHashMap<>();
+        Set<String> words = new LinkedHashSet<>();
```
Comment (minborg) [inline src/test/java/net/openhft/chronicle/wire/NestedMapsTest.java:258]: Can't mutable objects be final?

**Comment by peter-lawrey** [comment]
It's important to recognise how Java Serialization does this currently, as we can't change that behaviour. Java Serialization doesn't produce a warning when setting final fields, nor does it expect you to make a field that can be set non-final

**Pull request closed without merging**